sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,658 @@
+also be used to name the variable in SPIR-V, so it's better for debuggability
+if the string matches the mangled name for variables with external linkage.
+This is not possible, though, for variables with internal linkage because the
+mangled name is not unique in this case.  For these variables, we use the
+mangled name and append a unique suffix.
+
+Each device global variable that has the `device_image_scope` property is also
+decorated with the `sycl-device-global-image-life` attribute.
+
+Note that language rules ensure that `device_global` variables are always
+declared at namespace scope (i.e. a global variable), and LLVM IR [allows
+attributes to be attached to global variables][2].
+
+[2]: <https://llvm.org/docs/LangRef.html#global-attributes>
+
+### Changes to the DPC++ front-end
+
+The device compiler front-end is changed in two ways: it generates new content
+in both the integration header and the integration footer, and it adds the
+`sycl-unique-id` and `sycl-device-global-image-life` attributes to the IR
+definitions of `device_global` variables as defined above.  These two tasks are
+related because the integration footer contains the same string that is stored
+in the `sycl-unique-id` attribute.
+
+**NOTE**: See also the "Unresolved issues" section at the bottom of this
+document for other changes that are needed in the front-end.
+
+#### New content in the integration header and footer
+
+New content in the integration header and footer provides a mapping from the
+host address of each device global variable to the unique string for that
+variable.  To illustrate, consider a translation unit that defines two
+`device_global` variables:
+
+```
+#include <sycl/sycl.hpp>
+
+static sycl::device_global<int> Foo;
+namespace inner {
+  sycl::device_global<double[2]> Bar;
+} // namespace inner
+
+// ...
+```
+
+The corresponding integration header defines a namespace scope variable of
+class type whose sole purpose is to run its constructor before the
+application's `main()` function:
+
+```
+namespace sycl::detail {
+namespace {
+
+class __sycl_device_global_registration {
+ public:
+  __sycl_device_global_registration() noexcept;
+};
+__sycl_device_global_registration __sycl_device_global_registerer;
+
+} // namepsace (unnamed)
+} // namespace sycl::detail
+```
+
+The integration footer contains the definition of the constructor, which calls
+a function in the DPC++ runtime with the following information for each device
+global variable that is defined in the translation unit:
+
+* The (host) address of the variable.
+* The variable's string from the `sycl-unique-id` attribute.
+* The size (in bytes) of the underlying `T` type for the variable.
+* A boolean telling whether the variable is decorated with the
+  `device_image_scope` property.
+
+```
+namespace sycl::detail {
+namespace {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  device_global_map::add(&::Foo,
```

Comment:
I can't find any detailed description of that function, @romanovvlad, @s-kanaev, is that ok with you?

Am I right that runtime will likely have to maintain two mappings: one from `device_global` address to metadata and another one from string identifier of `device_global` to metadata (including address of `device_global`)? The first one would be used in implementation of copy functions and the second one will be used during global variables initialization

Reply:
Yes.
Yes, + a mapping from a device to USM ptr.